W-10587214 - Transform to SQL and server-side pagination explanations - CP (#82)
mulesoft/docs-apikit: modules/ROOT/pages/creating-an-odatav4-api-with-apikit.adoc

== Use The Transform to SQL Operation

> APIkit for OData 4 contains a `Transform to SQL Select` operation to transform the input OData `GET` request to a SQL SELECT query. The operation is useful when combined with the `select` operation of the Database Connector. The output payload is a prepared statement and the attributes are a map of prepared statement parameters. It has been designed to work with the Mule DB connector to help the user expose databases with OData 4 easily.

Reviewer: -"**It** has been designed" > **This operation** has been designed... 
Not completely clear what **it** refers to, please confirm if this suggestion is accurate. 

-to help the user expose databases with OData 4 easily > to simplify the process of exposing databases with OData 4.
This is just a suggestion, consider if you prefer this change or the original.

Author: The first comment committed! ;) Thanks!

image::transform-to-sql-transformer.png["Transform to SQL select element and tab are highlighted."]

image::add-db-transform-to-sql.png["Select element and tab are highlighted, as well as the attributes.queryparameters in the input parameters field."]

image::serialize-transform-to-sql.png["Serialized entity element and tab are highlighted, as well as the inbound content in the General section."]

> On the one hand, APIkit currently supports the following elements:

Reviewer: On the one hand > Try to avoid idiomatic expressions which can cause issues during localization. In this case, just removing it is enough, no need to replace it: 
APIkit currently supports the following elements:

Author: OK! Thanks for the tip!

* System query options:
** $filter
** $select
** $search
** $orderBy
** $skip
** $top
* Single entity
* Entity relation
* Server-side pagination

> On the other hand, APIkit does not currently support the following elements:

Reviewer: Same comment, leave only: APIkit does not currently support the following elements:

Author: OK! Thanks for the tip!

* Parameter Aliases
* Built-in Filter Operations:
** has
** divby 
* Built-in Query Functions, all but:
** contains
** startswith
** endwith
* $apply
* $compute
* $expand

> === How does transform to SQL select work in OData 4?

Reviewer: === How Does Transform to SQL Select Work in OData 4?

Fix capitalization

Author: Capitalization fixed! ;) Thanks!

> The OData protocol defines a series of conventions that allow clients to query identified resources using URLs. The list of queryable entities, their fields, and the relationships between them are defined in a data model using simple HTTP messages.

Reviewer: that **allow** clients > that **enable** clients (per style ref.)

Author: OK! Thanks!

> As OData shares some similarities with Java database connectivity (JDBC), the `transform-to-sql-select` operation takes advantage of such similarities to help the user create SQL queries for those resources. Then, the query is returned at the payload along with the query parameters at the attributes.

Reviewer: "Then, the query is returned at the payload along with the query parameters at the attributes."
This sentence is unclear, please address the following points: 
- "the query is returned" > who or what returns the query? Avoid passive voice by specifying **what** returns the query
- "returned at the payload" > I can't understand this, does this mean "returned **to** the payload?"
- "query parameters at the attributes" > also confusing, was this intended to be "**as** the attributes" or "**and** the attributes" ? 
Please check with devs to clarify these points.

Author: Requesting APIkit team assistance to get the answers needed to avoid ambiguity in these paragraphs.

Reviewer 2: `transform-to-sql` is an operation that takes a description of a given API request as input and outputs a SQL query (as payload) and a "parameter -> value" mapping (as attributes).

Author: Hi, APIkit team!
Please, can you further explain this:
"query parameters at the attributes".
Is this intended to be "as the attributes" or "and the attributes"?
Taking into account Ignacio's comment, it was meant to say "as" (typo error).
Apart from that, please, can you take a look at line 143 ("As OData shares... with the query parameters at the attributes"). Is the current explanation accurate?
Thanks for your support!

Reviewer 3: The explanation is accurate but the last sentence sounds disconnected and out of context, it might be rephrased. As @iglosiggio mentioned, the idea is to explain the structure of the resulting Mule message of the operation (see [mule message documentation](https://docs.mulesoft.com/mule-runtime/4.4/about-mule-message) as reference)

Reviewer 3: The resulting Mule message ends up being a payload with a prepared statement of the SQL `SELECT` sentence as plain text, along with a map of attributes where keys are the name of the variables in the prepared statement and values the actual values of each variable.

Reviewer 2: If you look at the output of the operation you will see that the resulting Mule Message has a modified `Payload` and `Attributes` (typed accordingly... well, I think `Payload` should be typed as an `String` but nvm).
![image](https://user-images.githubusercontent.com/1154680/170129210-a66c951c-dfe4-4e26-b666-f204199a5441.png)

I don't know what's the best wording to express this. But the basic idea is that the payload of the Mule Message outputted by this operation _is_ the query and the attributes _is_ the query parameters mapping.

Does this help?

Author: Resolved over chat! ;)

> As OData shares some similarities with Java database connectivity (JDBC), the `transform-to-sql-select` operation takes advantage of such similarities to help the user create SQL queries for those resources. The `transform-to-sql-select` operation returns the query to the payload along with the query parameters as the attributes.

Reviewer 3: ```suggestion
As OData shares some similarities with Java database connectivity (JDBC), the `transform-to-sql-select` operation takes advantage of such similarities to help the user create SQL queries for those resources. Therefore, the resulting message of the operation is a prepared statement with the SQL SELECT sentence as the payload along with the parameters name and values map as the attributes.
```
"the `transform-to-sql-select` operation" gets duplicated in the last sentence. I would rephrase this as suggested. Wjat do you think @iglosiggio ?

Author: Resolved over chat! ;)

> === How do I implement the transform to SQL select operation using APIkit for OData 4?

Reviewer: === How Do I Implement the Transform to SQL Select Operation Using APIkit for OData 4?

Author: Title changed! Thanks!

> <apikit-odata:transform-to-sql-select> is the name of the XML element that represents this operation, which has the following parameters: 

Reviewer 3: Please, add back ticks to this XML, as suggested before for a different section

Author: Right you are! Thanks!

> <apikit-odata:transform-to-sql-select> is the name of the XML element that represents this operation, which has the following parameters: 

Reviewer 3: Please, add back ticks to this XML, as suggested before for a different section

Author: Right you are! Thanks!

|====
| *Parameter name* | *Description*
| Display name | This parameter configures the operation name.
| Module configuration | Name of the related-module configuration.
| Entity set name | The name of the entity set that you are trying to fetch #[attributes.entitySetName].
| Page size | A number that defines the amount of results per page during server-side-pagination.
| Entity type name | A string obtained from the OData request using #[attributes.odataRequestAttributes], which is specified by the entity data model.
| Entity type fields | An array of strings obtained from the OData request using #[attributes.odataRequestAttributes], which provides the list of property fields from the entity type.
| Entity type keys | An object obtained from the OData request using #[attributes.odataRequestAttributes], which provides the map of entity key names and values.
| System query options | Parameter obtained from the OData request using #[attributes.odataRequestAttributes].
| Search transformation enabled | Parameter to define the behavior of the transformation.
| Strict search term match | Parameter to define the behavior of the transformation.
|====

> === $filter expressions in OData 4 

Reviewer: === $filter Expressions in OData 4

Author: Title changed! Thanks!

> ​​The `$filter` expression in OData uses comparison, logical, arithmetic, and grouping operators. You can access the list operators with their corresponding meaning at http://docs.oasis-open.org/odata/odata/v4.01/cs01/part1-protocol/odata-v4.01-cs01-part1-protocol.html#sec_BuiltinFilterOperations[the OData4 Specification]. As mentioned before, APIkit does not support `has` and `divby` operators.

Reviewer: - [the OData4 Specification] > [the OData4 Specification^]
Add **^** when it is an external link so it displays an icon next to the URL. 
- As mentioned before, > Remove this, start the sentence directly with "APIkit does not...

Author: Thanks for the ^ tip! 
Phrase removed!
;)

> === Supported expressions by transform to SQL select

Reviewer: === Expressions Supported by Transform to SQL Select 
Suggesting a minor rewording

Author: You're right! Thanks!

> OData supports a variety of expressions for filtering that you can find in the [specification](http://docs.oasis-open.org/odata/odata/v4.01/cs01/part1-protocol/odata-v4.01-cs01-part1-protocol.html#sec_BuiltinFilterOperations). `transform-to-sql-select` operation supports the following subset of unary and binary operators, along with the following method operators.

Reviewer: - Please use the same syntax as before in this URL link: _https://docs.oasis-open.org...[specification^]_
- `transform-to-sql-select` operation > **The** `transform-to-sql-select` operation

Author: Thanks!

> OData supports a variety of expressions for filtering that you can find in the [specification^](http://docs.oasis-open.org/odata/odata/v4.01/cs01/part1-protocol/odata-v4.01-cs01-part1-protocol.html#sec_BuiltinFilterOperations). The `transform-to-sql-select` operation supports the following subset of unary and binary operators, along with the following method operators.

Reviewer 3: is the link to the specification well formatted? It has a different format as the one in the previous section.

Author: Good catch! Resolved!

> OData supports a variety of expressions for filtering that you can find in the [specification^](http://docs.oasis-open.org/odata/odata/v4.01/cs01/part1-protocol/odata-v4.01-cs01-part1-protocol.html#sec_BuiltinFilterOperations). The `transform-to-sql-select` operation supports the following subset of unary and binary operators, along with the following method operators.

Reviewer 3: is the link to the specification well formatted? It has a different format as the one in the previous section.

Author: Good catch! Resolved!

==== Binary Operators in OData and in SQL

|====
| *OData Operator* | *SQL Operator*
| add | +
| and | AND
| div | /
| eq | =
| ge | >=
| gt | >
| le | =<
| lt | <
| mod | %
| mul | *
| ne | <>
| or | OR
| sub | -
|====

==== Unary Operators in OData and in SQL

|====
| *OData Operator* | *SQL Operator*
| not | NOT 
| - | - 
|====

==== Method Operators in OData and in SQL

|====
| *OData Operator* | *SQL Operator*
| contains | LIKE %value%
| startswith | LIKE value%
| endswith | LIKE %value
|====

> === $search support

Reviewer: === $search Support

Author: Thanks!

> The operation builds a “where” clause based on the `$search` expression provided by the system query option. The implementation translates the search expression to select all the elements that have properties values that match the condition. By default, a term matches if the field value “contains” the term, but it will not match if it is set to be “strict” by configuration. The following parameters define the `$search` feature:

Reviewer: - “where” > `where` (use backticks ` if this is talking about the WHERE clause in SQL)
- that have **properties** values > that have **property** values
- but it **will not** match > but it **does not** match (avoid will, use present)
- if it is set **to be** “strict” **by** configuration. > if it is set **as** "strict" **in the** configuration.

Author: Great catches! Thanks!

The following parameters define the `$search` feature:

* Search transformation enabled: it enables (true) or disables (false) the transformation support for this option.

> * Strict search term match: it defines if the match should be strict (true) or not strict (false), meaning that the field value should exactly match the term(s) of the search expression.

Reviewer 3: These are parameters of the operation too. I would keep the format similar to the way that we described the rest of the parameters in line #140, using a table. What do you think about that?

Author: Good catch! Change done! ;)

> | Strict search term match | It defines if the match should be strict (true) or not strict (false), meaning that the field value should exactly match the term(s) of the search expression.

Reviewer: Suggested rewordings: 
- | When set to `true`, enables transformation support for this option. 
- I When set to `true`, specifies that there is a match only if the field value contains exactly the same terms as the search expression.
(please confirm with devs if these suggestions are accurate)

Reviewer 2: The suggestions are accurate, maybe it would be nice to explain what we mean by "exactly the same" in a better way.

Suppose you can have a record:
```
Person
=====
Name: Ignacio
Surname: Losiggio
Nickname: Mega
```

Only the terms `"Ignacio"`, `"Losiggio"` or `"Mega"` will match that record under a strict `$search`.
On the other hand, for a non-strict `$search` any substring of either `"Ignacio"`, `"Losiggio"` or `"Mega"` will match (things like `"naci"`, `"iggi"`, `"M"`).

Author: Requesting assistance from APIkit team.

Author: Table content has been changed! ;) 
Thanks!

== Use The Transform to SQL Operation

APIkit for OData 4 contains a `Transform to SQL Select` operation to transform the input OData `GET` request to a SQL SELECT query. The operation is useful when combined with the `select` operation of the Database Connector. The output payload is a prepared statement and the attributes are a map of prepared statement parameters. It has been designed to work with the Mule DB connector to help the user expose databases with OData 4 easily.

image::transform-to-sql-transformer.png["Transform to SQL select element and tab are highlighted."]

image::add-db-transform-to-sql.png["Select element and tab are highlighted, as well as the attributes.queryparameters in the input parameters field."]

image::serialize-transform-to-sql.png["Serialized entity element and tab are highlighted, as well as the inbound content in the General section."]

On the one hand, APIkit currently supports the following elements:

* System query options:
** $filter
** $select
** $search
** $orderBy
** $skip
** $top
* Single entity
* Entity relation
* Server-side pagination

On the other hand, APIkit does not currently support the following elements:

* Parameter Aliases

> * Built-in Filter Operations (has, divby) 

Reviewer 4: Is clear enough that the only not supported Built-in Filter Operations are has an divby?

Reviewer 3: Yeah, it is not clear enough I think. I would go with the suggestion done by @iglosiggio below.

Author: For me, it was not clear enough so I made some changes to make it more clear.
Thanks for the comments!

* $apply
* $compute
* $expand

=== How does transform to SQL select work in OData 4?

The OData protocol defines a series of conventions that allow clients to query identified resources using URLs. The list of queryable entities, their fields, and the relationships between them are defined in a data model using simple HTTP messages.
As OData shares some similarities with Java database connectivity (JDBC), the `transform-to-sql-select` operation takes advantage of such similarities to help the user create SQL queries for those resources. Then, the query is returned at the payload along with the query parameters at the attributes.

=== How do I implement the transform to SQL select operation using APIkit for OData 4?

<apikit-odata:transform-to-sql-select> is the name of the XML element that represents this operation, which has the following parameters: 

|====
| *Parameter name* | *Description*
| Display name | This parameter configures the operation name.
| Module configuration | Name of the related-module configuration.
| Entity set name | The name of the entity set that you are trying to fetch #[attributes.entitySetName].

> | Page size | A number that defines the amount of pages in server-side-pagination.

Reviewer 2: Is this correct?

The name of the parameter suggests this interpretation. I don't remember what we have implemented (cc @juanchib @JuanAller @dcourtil)
```suggestion
| Page size | A number that defines the amount of results per page during server-side-pagination.
```

Reviewer 3: Good catch and suggestion!

Author: Thanks for the suggestion!

| Entity type name | A string obtained from the OData request using #[attributes.odataRequestAttributes], which is specified by the entity data model.
| Entity type fields | An array of strings obtained from the OData request using #[attributes.odataRequestAttributes], which provides the list of property fields from the entity type.
| Entity type keys | An object obtained from the OData request using #[attributes.odataRequestAttributes], which provides the map of entity key names and values.
| System query options | Parameter obtained from the OData request using #[attributes.odataRequestAttributes].
| Search transformation enabled | Parameter to define the behavior of the transformation.
| Strict search term match | Parameter to define the behavior of the transformation.
|====

=== $filter expressions in OData 4 

> ​​The `$filter` expression in OData uses comparison, logical, arithmetic, and grouping operators.
> 
> ==== Comparison Operators
> 
> |====
> | *Operator* | *Description* | *Example* 
> | eq | Equal | Address/City eq 'Redmond'
> | ne | Not equal | Address/City ne 'London'
> | gt | Greater than | Price gt 20
> | ge | Greater than or equal to | Price ge 10
> | lt | Less than | Price lt 20
> | le | Less than or equal to | Price le 100
> | in | Is a member of | Address/City in ('Redmond', 'London')
> |====
> 
> ==== Logical Operators
> 
> |====
> | *Operator* | *Description* | *Example*
> | and | Logical and | Price le 200 and Price gt 3.5
> | or | Logical or | Price le 3.5 or Price gt 200
> | not | Logical negation | not endswith(Description,'milk')
> |====
> 
> ==== Arithmetic Operators
> 
> |====
> | *Operator* | *Description* | *Example*
> | add | Addition | Price add 5 gt 10
> | sub | Subtraction | Price sub 5 gt 10
> | mul | Multiplication | Price mul 2 gt 2000
> | div | Division | Price div 2 gt 4
> | mod | Modulo | Price mod 2 eq 0
> |====
> 
> ===== Grouping Operators
> 
> |====
> | *Operator* | *Description* | *Example*
> | ( ) | Precedence grouping | (Price sub 5) gt 10
> |====

Reviewer 2: As discussed on Slack, it _might_ be a good idea to just direct the Mule dev to the original source of this table:

```suggestion
​​The `$filter` expression in OData uses comparison, logical, arithmetic, and grouping operators. You can see the list operators with their corresponding meaning at http://docs.oasis-open.org/odata/odata/v4.01/cs01/part1-protocol/odata-v4.01-cs01-part1-protocol.html#sec_BuiltinFilterOperations[the OData4 Specification]. The only unsupported operators are `has` and `divby`.
```

(Obviously take the suggested wording just as an idea of the information we want to provide, it's not something set in stone, lol).

Author: Changed accepted and paragraph reworded.

> |====

Reviewer 3: I would get rid of this and instead point it out to the [ODATA spec](http://docs.oasis-open.org/odata/odata/v4.01/cs01/part1-protocol/odata-v4.01-cs01-part1-protocol.html#sec_BuiltinFilterOperations) in the below section

=== Supported expressions by transform to SQL select

> `transform-to-sql-select` operation supports the following unary and binary operators, along with the following method operators.

Reviewer 3: ```suggestion
OData supports a variety of expressions for filtering that you can find in the [specification](http://docs.oasis-open.org/odata/odata/v4.01/cs01/part1-protocol/odata-v4.01-cs01-part1-protocol.html#sec_BuiltinFilterOperations). `transform-to-sql-select` operation supports the following subset of unary and binary operators, along with the following method operators.
```

image::serialize-transform-to-sql.png["Serialized entity element and tab are highlighted, as well as the inbound content in the General section."]

=== What is transform to SQL select in OData 4?

> The `transform-to-sql-select` is an operation that allows the user to transform an OData query request into a SQL select statement. It has been designed to work with the Mule DB connector to help the user expose databases with OData 4 easily.

Reviewer 3: This says pretty much the same as the introduction that already exists, is it worth to merge both or keep only one in some way? or do you think that is worth to refresh what the operation is meant to do in this new section?

Author: You may be right that both paragraphs may express the same idea with different words.
But the first paragraph is meant to describe the transform to SQL operation, and the second paragraph is meant to engage the reader to use such an operation by explaining that it works with the Mule DB connector and providing the elements it supports.
Nevertheless, we may merge both things and change the article layout to:

Use The Transform to SQL Operation
APIkit for OData 4 contains a Transform to SQL Select operation to transform the input OData GET request to a SQL SELECT query. The operation is useful when combined with the select operation of the Database Connector. The output payload is a prepared statement and the attributes are a map of prepared statement parameters. It has been designed to work with the Mule DB connector to help the user expose databases with OData 4 easily.
Image 1 
Image 2
Image 3
On the one hand, APIkit currently supports the following elements:
Etc.

Do you think this change would improve this article?

=== Server-Side Pagination in APIkit for OData 4

Server-side pagination is a way of controlling a subset of data requests that were fetched from a client-side server using a fixed `pageSize` parameter. It divides the response data set into discrete pages to improve data readability. The server defines the page size, which is the maximum number of records returned by the server in a request.

> ===== How server-side-pagination works in OData 4

Reviewer: ===== How Server-Side-Pagination Works in OData 4 
(unless server-side-pagination is an XML element or code, in which case it stays lowercase)

Author: Title changed! Thanks!

> OData defines a `@odata.nextLink` to indicate that a response is only a subset of the requested collection of entities. As such, `@odata.nextLink` contains a URL that allows the client to retrieve the next subset of the requested collection.

Reviewer: allows > enables

Author: Thanks!

> Apart from that, the URL contains a query parameter `$skiptoken` that indicates the server where the next page starts.

Reviewer: Remove _Apart from that,_ and leave this sentence next to the previous one.

Author: Thanks!

> ==== How to implement this feature in APIkit for OData 4

Reviewer: ==== How to Implement This Feature in APIkit for OData 4

Author: Title changed! Thanks!

> <apikit-odata:request-entity-collection-listener> allows you to define a set of fields in the *Responses* tab, in Anypoint Studio, to implement the server-side pagination.

Reviewer: - <apikit-odata:request-entity-collection-listener> allows > `<apikit-odata:request-entity-collection-listener>` enables... (add backticks ` to XML elements )

Author: Requesting assistance from APIkit team to learn which are the XML elements.

Reviewer 2: `apikit-odata:request-entity-listener`, `apikit-odata:request-entity-collection-listener`, `apikit-odata:serialize-entity-collection`, `apikit-odata:serialize-entity`, `apikit-odata:transform-to-sql-select`, `apikit-odata:route`, and `apikit-odata:config`

Reviewer 2: Here is a full list (of at least everything we have a test for):

```console
# Command to find all instances of apikit-odata:SOMETHING 
> git grep -oh apikit-odata:[^>/ ]\* | sort | uniq
apikit-odata:collection-success-response
apikit-odata:config
apikit-odata:entity-collection-success-response
apikit-odata:entity-success-response
apikit-odata:flow-response-metadata-properties
apikit-odata:flow-response-metadata-property
apikit-odata:http-request-parameters
apikit-odata:inbound-content
apikit-odata:raw-entity-collection-success-response
apikit-odata:raw-entity-success-response
apikit-odata:request-entity-collection-listener
apikit-odata:request-entity-listener
apikit-odata:request-parameters
apikit-odata:route
apikit-odata:serialize-entity
apikit-odata:serialize-entity-collection
apikit-odata:success-response
apikit-odata:transform-to-sql-select
```

Author: Change made! Thanks!

image::pagesize.png[alt text:"pageSize set to ${service.orders.pageSize} in the On Entity Collection Request tab."]

> Once populated, APIkit creates the XML representation of the previous fields.

Reviewer: - Once populated > After it is populated(Once also means "one time", this change helps avoid localization issues)

Author: OK! Thanks!

image:xmlrepresentation.png[alt text:"XML representation of the pageSize mentioned before."]

> ==== Page size or Token fields for server-side pagination

Reviewer: ==== Page Size or Token Fields for Server-Side Pagination

Author: Title changed! Thanks!

> `Page size` is a fixed number that represents the maximum number of records the request-entity-collection listener can retrieve for each request.

Reviewer: `request-entity-collection` (if this is an XML element)

Author: Thanks!

> In case you cannot use a numeric pagination, you can set a token in the `Token field`, which the developer will generate in the corresponding flow.

Reviewer: Who is the developer here? Is the user configuring the token so a developer on their team can later generate it in the flow? If this is correct, please consider the following rewording: 

In case you cannot use a numeric pagination, set a token in the `Token field` so a developer can generate it in the corresponding flow. 
(please check with devs for accuracy before making changes)

Reviewer 2: Yup.

The `Token field` encodes pagination information as a string. A simple way to do it is to encode the "page number" or the "offset from the start of the list" but in principle it might be any string the Mule dev can use to know which entities to retrieve.

For example on something sorted by time (where each record can be uniquely identified by the time attached to it) the token can be used to retrieve K of records "since" a given time.

The semantic meaning of the `Token field` should not be relevant to the API consumer. The should see these strings only as a way to ask for the next page.

Author: Thanks for the contribution! :)

> APIkit for OData will include this token value in the `@odata.nextLink` URL with a query parameter `$skiptoken=<Token_Value>`.

Reviewer: will include > includes

Author: "Will" removed!
Thanks!

=== Server-Side Pagination in APIkit for OData 4

> Server-side pagination is a way of controlling a subset of data requests that were fetched from a client-side server using a fix `pageSize` parameter. It divides the response data set into discrete pages to improve data readability. The server defines the page size, which is the maximum number of records returned by the server in a request.

Reviewer 2: Shouldn't this be "Using a fixed `pageSize` parameter"?

Looks like a typo but I might be wrong.

Reviewer 5: 👍
